packages/astro/src/manifest/virtual-module.ts

```diff
@@ -0,0 +1,111 @@
+import type { Plugin } from 'vite';
+import { CantUseManifestModule, ForbiddenManifestModule } from '../core/errors/errors-data.js';
+import { AstroError } from '../core/errors/index.js';
+import type { Logger } from '../core/logger/core.js';
+import type { AstroSettings } from '../types/astro.js';
+import type {
+	AstroConfig,
+	ClientConfigSerialized,
+	ServerConfigSerialized,
+} from '../types/public/index.js';
+
+const VIRTUAL_MODULES_IDS = {
+	client: 'astro:manifest/client',
+	server: 'astro:manifest/server',
+};
+
+const VIRTUAL_MODULES_IDS_SET = new Set(Object.values(VIRTUAL_MODULES_IDS));
+
+function resolveVirtualModuleId<T extends string>(id: T): `\0${T}` {
+	return `\0${id}`;
+}
+
+export default function virtualModulePlugin({
+	settings,
+	logger,
+}: { settings: AstroSettings; logger: Logger }): Plugin {
+	return {
+		enforce: 'pre',
+		name: 'astro-manifest-plugin',
+		resolveId(id) {
+			// Resolve the virtual module
+			if (VIRTUAL_MODULES_IDS_SET.has(id)) {
+				return resolveVirtualModuleId(id);
+			}
+		},
+		load(id, opts) {
+			// client
+			if (id === resolveVirtualModuleId(VIRTUAL_MODULES_IDS.client)) {
+				if (!settings.config.experimental.serializeManifest) {
+					throw new AstroError({
+						...CantUseManifestModule,
+						message: CantUseManifestModule.message(VIRTUAL_MODULES_IDS.client),
+					});
+				}
+				// There's nothing wrong about using `/client` on the server
```

Comment:
What would be the use case for this?

Reply:
Yeah I was also thinking about this during lunch. For `astro:env` it makes sense not to throw an error when accessing client vars on the server because you can't get them through the server module. However here, I don't see a good reason not to use the server module. I think in dev we could serve the module and log the warning, and throw during the build

Reply:
I'm actually thinking about changing the approach, and have two modules that return two different sets of information, instead of copy them.

For example, the i18n section of the config to be exported by `/client` because it's safe to have in the browser and not by `/server`, so we don't duplicate information when we build the app.

Reply:
I wonder if we should keep data server side to be honest

Reply:
A lot of this will be useful in the client, so I do think it should be available

Reply:
> I wonder if we should only keep data server side to be honest

Routing information is very useful to the client. If a SPA uses react-i18n, with module they have access to the locales